### DISCUSSION TITLE
Preprocessor Directive Comments/Helpers

### DISCUSSION THREAD

**Discussion by michael-hawker** [dotnet/csharplang#6112] [Language Ideas]
## Background

Currently, we can add explicit warnings or errors with preprocessor directives:

```cs
#error Deprecated code in this method.
#warning Deprecated code in this method.
```

[See doc section here.](https://docs.microsoft.com/dotnet/csharp/language-reference/preprocessor-directives#error-and-warning-information)

## Problem

However, sometimes I want to decorate the `#if` conditional with info about why it's there (vs. the code that it's conditioning), or maybe it's blocking out a specific platform, but I want to track be able to go back to fix it later for others.

There's not a good way to track these things today. It'd be nice to have a couple of new options to either track important comments or things that should be addressed in the future.

Also see this related [Stack Overflow issue](https://stackoverflow.com/questions/51609098/writing-comments-for-c-sharp-pre-processors) from the past.

## `#comment` conditional

This is completely ignored and just used as a comment to help explain other directives.

```cs
#comment we currently can't run on platform X because see https://someurl.com/somelink
#if __SOME_PLATFORM__
```

Could be handy if this could be shortcut with a `##` like so:

```cs
## we currently can't run on platform X because see https://someurl.com/somelink
#if __SOME_PLATFORM__
```

## `#info` conditional

This would emit during the build output a message at the `info` log level that would not be a warning or an error. It would allow some context for certain things and potentially help folks with debugging complex build parts or knowing which conditionals are on.

```cs
#info message here
```

```cs
#info this message appears always
#if DEBUG
#info THIS IS A DEBUG BUILD
#endif
```

## `#todo` conditional

This would be used to track more traditional `// TODO:` comments in a compiler detectable way to be used by future tooling and VS improvements (like some [TODO window extensions](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) which pull out of comments and text). Could be handy for these to show up in GitHub or DevOps to auto-create or prompt creating issues, etc...

```cs
#todo 42 we should optimize this code later when we know things are working
```

```cs
#todo <optional number> message
```

The number would be an id to differentiate between different tasks (or link related tasks across code), could be used by automated tools for tracking/aggregating items.

**Comment by CyrusNajmabadi** [comment]
> #comment conditional

Can't this just be a normal comment.  e.g.:

```c#
// comment we currently can't run on platform X because see https://someurl.com/somelink
#if __SOME_PLATFORM__
```

> This would be used to track more traditional // TODO: comments in a compiler detectable way to be used by future tooling and VS improvements (like some [TODO window extensions](https://marketplace.visualstudio.com/items?itemName=Gruntfuggly.todo-tree) which pull out of comments and text). Could be handy for these to show up in GitHub or DevOps to auto-create or prompt creating issues, etc...

Why do we need a compiler-detectable way?  We already have a TODO window in VS and it already can find comments marked with whatever identifiers you specify you want to be found.  e.g.:

![image](https://user-images.githubusercontent.com/4564579/167731975-9ba706fc-8945-450f-baa6-b68ef316d0b4.png)

**Reply by HaloFour** [reply]
That feels like something an analyzer could do too, if you wanted compiler diagnostics emitted from your build.
